Add Physical Server refresh

MaysaMacedo opened the pull request:
This PR contains the trigger for the action of refreshing a Physical Server.

![image](https://user-images.githubusercontent.com/5839808/32189644-251df12c-bd8a-11e7-8a64-2a3e8f211368.png)

![image](https://user-images.githubusercontent.com/5839808/32221678-36ce2168-be15-11e7-8be5-0f6b9ce8368a.png)


dependes on [#16344](https://github.com/ManageIQ/manageiq/pull/16344)

MaysaMacedo commented:
@AparnaKarve @agrare @juliancheal Could you take a look on this PR?

MaysaMacedo commented:
@miq-bot add_label enhancement, providers/physical

mzazrivec commented:
@MaysaMacedo please remove the `pending core` label once the backend PR is merged, so that
this PR can be reviewed / merged.

MaysaMacedo commented:
@miq-bot add_label gaprindashvili/yes

agrare commented:
@miq-bot remove-label pending core

AndreyMenezes commented:
@AparnaKarve @agrare @chessbyte Could you take a look on this PR?
The problem with the code climate is that the codes was similar, I don't see how do it from another way. Suggestions?

AndreyMenezes commented on app/controllers/mixins/actions/physical_server_actions/misc.rb:
```diff
@@ -0,0 +1,66 @@
+        end
+
+        # Refresh all selected or single displayed physical servers(s)
+        def refreshphysicalservers
```
Why this particular method didn't use underscore in the name?

MaysaMacedo replied:
Just followed the pattern for the refreshs methods.

martinpovolny replied:
please don't that is old code not yet refactored, please, make it uderscores

rodneyhbrown7 commented:
@miq-bot assign @dclarizio

AparnaKarve commented:
@MaysaMacedo Did you use the `refreshhosts` method as a reference for `refreshphysicalservers`?
Both methods seem so identical - and the only difference is the model name - `Host` Vs `PhysicalServer`

If both `Hosts` and `PhysicaServers` could use one common method, that would be great.
However, that would mean a little bit of refactoring.

@martinpovolny Do you think refactoring would be a good idea right now? 
Or leave the current `refreshphysicalservers` implementation as-is for now, and consider refactoring after Gaprindashvili has stabilized?

MaysaMacedo commented:
@AparnaKarve I did use the refreshhosts as a reference, since it does provide the same functionality for Physical Servers and triggers the action of refresh based on the same conditions.

martinpovolny commented on app/controllers/mixins/actions/physical_server_actions/misc.rb:
```diff
@@ -0,0 +1,66 @@
+          physical_servers, _physical_servers_out_region = filter_ids_in_region(physical_servers, _("PhysicalServer"))
+          return if physical_servers.empty?
+
+          display_name ||= task
```
why this line?

jsut make `display_name` obligatory 

if you don't and the task ever get's into display_name then in `process_physical_servers_refresh` the flash message will not be properly internationalized

you probably followed some bad pattern again, but seems you don't need this here.

martinpovolny commented:
I see that you are adding new functionality following the old code. This means that this functionality will not be available through the API.

The best way to move forward is to go API first. Meaning you would create the API for the functionality that you need and then you would create a UI that would use the new API.

There's a nice guide on creating the APIs: http://manageiq.org/blog/2017/10/api-contributors-guide/

MaysaMacedo commented:
@martinpovolny Thanks. I've updated with the modifications you suggested and added support for refresh in the API(https://github.com/ManageIQ/manageiq-api/pull/189).

AndreyMenezes commented:
@martinpovolny @AparnaKarve The dependency was merged! Could you take a look on this PR?

chargio commented:
@martinpovolny @AparnaKarve Anything missing in this PR?

Thanks,
Sergio

AparnaKarve commented:
@MaysaMacedo @AndreyMenezes After ManageIQ/manageiq-api#189 merge, you should be able to use this API for Physical Server refresh called directly from the toolbar - which also means, you do not need the implementation in this PR.

Please refer to the Angular JS code in app/assets/javascripts/components/physical_infrastructures/physical-server-toolbar.js to get an idea on how to call the API to perform an action on the Physical Servers.

PR https://github.com/ManageIQ/manageiq-ui-classic/pull/1380 should also be useful.
Thanks.

/cc @chargio

MaysaMacedo commented:
@AparnaKarve  Thank you. I've created another PR (#2843) for this feature and will close this one.
